barretenberg/cpp/pil/vm2/alu.pil

```diff
@@ -188,223 +322,193 @@ sel_mul_div_u128 { b_lo, constant_64 } in range_check.sel_alu { range_check.valu
+// which c_hi we choose. Our tracegen sets c_hi to 0 for FF values.
 #[ALU_MUL_NON_U128]
-sel_op_mul * IS_NOT_U128 * (1 - sel_tag_err) * (ia * ib - ic - (max_value + 1) * c_hi) = 0;
+sel_op_mul * IS_NOT_U128 * (1 - sel_err) * (ia * ib - ic - (max_value + 1) * c_hi) = 0;
 
 // MUL - u128
 
-pol commit sel_mul_u128;
-// sel_op_mul & sel_is_u128:
-sel_mul_u128 = sel_is_u128 * sel_op_mul;
-
-// Taken from vm1:
-// We express a, b in 64-bit slices: a = a_l + a_h * 2^64
-//                                   b = b_l + b_h * 2^64
-// => a * b = a_l * b_l + (a_h * b_l + a_l * b_h) * 2^64 + (a_h * b_h) * 2^128 = c_hi_full * 2^128 + c
-// => the 'top bits' are given by (c_hi_full - (a_h * b_h)) * 2^128
-// We can show for a 64 bit c_hi = c_hi_full - (a_h * b_h) % 2^64 that:
-// a_l * b_l + (a_h * b_l + a_l * b_h) * 2^64 = c_hi * 2^128 + c
-// Equivalently (cf = 0 if a_h & b_h = 0):
-// a * b_l + a_l * b_h * 2^64 = (cf * 2^64 + c_hi) * 2^128 + c
-// => no need for a_h in final relation
+// We express a, b in 64-bit slices: a = a_lo + a_hi * 2^64
+//                                   b = b_lo + b_hi * 2^64
+// => a * b = a_lo * b_lo + (a_hi * b_lo + a_lo * b_hi) * 2^64 + (a_hi * b_hi) * 2^128 = c_hi_full * 2^128 + c
+// for c_hi_full and c being 128-bit limbs. We can remove the term a_hi * b_hi * 2^128 when working modulo 2^128
+// and to avoid handling field overflows. We rewrite the relation as:
+// a * b_lo + a_lo * b_hi * 2^64 = (cf * 2^64 + c_hi) * 2^128 + c for a bit cf and c_hi < 2^64.
+// Namely, the left hand side is strictly less than 2 * 2^192.
+// c_hi can be computed explicitly as (c_hi_full - (a_hi * b_hi)) % 2^64 and
+// cf == 1 iff c_hi_full - (a_hi * b_hi) >= 2^64 equivalently cf == (c_hi_full - (a_hi * b_hi)) / 2^64.
 
 #[ALU_MUL_U128]
-sel_mul_u128 * (1 - sel_tag_err)
-    * (
+SEL_MUL_U128 * (
         ia * b_lo + a_lo * b_hi * TWO_POW_64            // a * b without the hi bits
         - ic                                            // c_lo
         - (max_value + 1) * (cf * TWO_POW_64 + c_hi)    // c_hi * 2^128 + (cf ? 2^192 : 0)
     ) = 0;
 
-// No need to range_check c_hi for cases other than u128 because we know a and b's size from the tags and have looked
-// up max_value. i.e. we cannot provide a malicious c, c_hi such that a + b - c_hi * 2^n = c passes for n < 128.
-// No need to range_check c_lo = ic because the memory write will ensure ic <= max_value.
-#[RANGE_CHECK_MUL_U128_C_HI]
-sel_mul_u128 { c_hi, constant_64 } in range_check.sel_alu { range_check.value, range_check.rng_chk_bits };
-
-// DIV
-
-sel_op_div * (1 - sel_op_div) = 0;
-
+// For FF values, this range check would not be required but it makes code simpler.
+pol commit sel_mul_no_err; // @boolean Follows from definition.
+sel_mul_no_err = (1 - sel_err) * sel_op_mul;
+#[RANGE_CHECK_MUL_C_HI]
+sel_mul_no_err { c_hi, constant_64 } in range_check.sel_alu { range_check.value, range_check.rng_chk_bits };
```

Comment:
Agree that it makes the code simpler but isn't it worth the overhead to gate the range check by non FF as well? i.e. `sel_mul_no_err = 1` if there is no err, if we are doing a `MUL`, and if we don't have an `FF`